src/main/java/com/google/cloud/tools/managedcloudsdk/install/SdkInstaller.java

```
@@ -0,0 +1,174 @@
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.managedcloudsdk.install;
+
+import com.google.cloud.tools.managedcloudsdk.MessageListener;
+import com.google.cloud.tools.managedcloudsdk.OsInfo;
+import com.google.cloud.tools.managedcloudsdk.Version;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import com.google.common.util.concurrent.MoreExecutors;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
+
+/** Install an SDK by downloading, extracting and if necessary installing. */
+public class SdkInstaller {
+
+  private final ListeningExecutorService executorService;
+  private final FileResourceProviderFactory fileResourceProviderFactory;
+  private final ExtractorFactory extractorFactory;
+  private final DownloaderFactory downloaderFactory;
+  private final InstallerFactory installerFactory;
+
+  /** Use {@link #newInstaller} to instantiate. */
+  SdkInstaller(
+      FileResourceProviderFactory fileResourceProviderFactory,
+      DownloaderFactory downloaderFactory,
+      ExtractorFactory extractorFactory,
+      InstallerFactory installerFactory,
+      ListeningExecutorService executorService) {
+    this.fileResourceProviderFactory = fileResourceProviderFactory;
+    this.executorService = executorService;
+    this.downloaderFactory = downloaderFactory;
+    this.extractorFactory = extractorFactory;
+    this.installerFactory = installerFactory;
+  }
+
+  /**
+   * Download a Cloud SDK asynchronously.
+   *
+   * @param messageListener a listener on installer output
+   * @return a path to the Cloud SDK home directory (wrapped in a listenable future)
+   */
+  public ListenableFuture<Path> downloadSdk(final MessageListener messageListener) {
```

Should there be something that makes sure this is only called once, since it assumes a running executorService?

Or maybe throw an error if the submit is rejected?

---

added in an illegalstateexception